Functions

joerivanarkel opened the pull request.

teusvanarkel commented on Common/EnvironmentVariables.cs:
```diff
@@ -7,13 +10,32 @@ namespace Common
 
     public class EnvironmentVariables : IEnvironmentVariables
```
Make singleton and read file once ( in constructor )

teusvanarkel commented on PirateInterpreter/Interpreters/InterpreterFactory.cs:
```diff
@@ -6,6 +8,10 @@ public BaseInterpreter GetInterpreter(INode node, ILogger logger)
+            case FunctionDeclarationNode functionDeclarationNode:
+                return new FunctionDeclarationInterpreter(functionDeclarationNode, this, logger);
+            case FunctionCallNode functionCallNode:
+                return new FunctionCallInterpreter(functionCallNode, this, logger, new StandardLibraryFactory());
```
Use DI for the StandardLibraryFactory

teusvanarkel commented on PirateInterpreter/StandardLibrary/IOLibrary.cs:
```diff
@@ -0,0 +1,23 @@
+
+namespace PirateInterpreter.StandardLibrary;
+
+public class IOLibrary
```
Pass the logger via the constructor instead of in the function

teusvanarkel commented on PirateInterpreter/StandardLibrary/IOLibrary.cs:
```diff
@@ -0,0 +1,23 @@
+
+public class IOLibrary
+{
+    public BaseValue Print(IList<BaseValue> parameters, ILogger logger)
```
Make method void or use Task

teusvanarkel commented on PirateInterpreter/StandardLibrary/IOLibrary.cs:
```diff
@@ -0,0 +1,23 @@
+        return null;
+    }
+
+    public BaseValue Read(IList<BaseValue> parameters, ILogger logger)
```
Make: public StringValue instead of BaseValue

teusvanarkel commented on PirateInterpreter/StandardLibrary/StandardLibraryFactory.cs:
```diff
@@ -0,0 +1,30 @@
+
+namespace PirateInterpreter.StandardLibrary;
+
+public class StandardLibraryFactory
```
Rename to better name...its not a factory

teusvanarkel commented on PirateInterpreter/StandardLibrary/StandardLibraryFactory.cs:
```diff
@@ -0,0 +1,30 @@
+
+public class StandardLibraryFactory
+{
+    public BaseValue GetFunction(string libraryname, string functionName, List<BaseValue> parameters, ILogger logger)
```
change BaseValue to StringValue

teusvanarkel commented on PirateInterpreter/Values/FunctionValue.cs:
```diff
@@ -0,0 +1,22 @@
+
+public class FunctionValue : BaseValue
+{
+    public IFunctionDeclarationNode functionDeclarationNode { get; set; }
```
first letter Capitalized

teusvanarkel commented on PirateParser/Node/FunctionDeclarationNode.cs:
```diff
@@ -0,0 +1,66 @@
+        ReturnType = returnType;
+        Statements = statements;
+        ReturnNode = returnNode;
+    }
```
call the first constructor with the other values

teusvanarkel commented on PirateParser/Parsers/FunctionDeclartionParser.cs:
```diff
@@ -0,0 +1,73 @@
+    {
+    }
+    // func name(parameters) : returntype { body }
+    public override (INode node, int index) CreateNode()
```
SingleResponsability

teusvanarkel commented on Shell/Commands/RunCommand.cs:
```diff
@@ -35,13 +35,23 @@ public override void Run(string[] arguments)
-                Console.WriteLine(item.Value.ToString());
-            }
-        }
+        // foreach (var item in interpreterResult)
```
remove commented

teusvanarkel reviewed (commented):
a few litle things

joerivanarkel commented:
**Summary**:
- Add functions to the language
    - New `FunctionDefinitionParser` parser and  new `FunctionCallNode`, `FunctionDefinitionNode`, `ParameterDefinitionNode` nodes
    - Added `FunctionCallInterpreter` and `FunctionDefinitionInterpreter`
    - Removed the automatic result printing, replaced by the new `IO.print()` function
- Started the Standard Libraries
    - Added a Function Call manager
    - Created the `IOLibrary`
- Moved `variables.json` into `/bin` folder
    - Template created upon start
- Update to .net 7
- Small Logger rework
